### DISCUSSION TITLE
Shader Outline

### DISCUSSION THREAD

**Discussion by peiyanlu** [iTwin/itwinjs-core#7046] [Q&A] [answered]
I want to add contour lines to the model through shaders, but I don't know much about shaders. Does iTwin support it

**Comment by pmconne** [comment]
You can configure a display styles [thematic settings](https://www.itwinjs.org/reference/core-common/symbology/thematicdisplay/) to produce [isolines](https://www.itwinjs.org/learning/display/thematicdisplay/).

![image](https://github.com/user-attachments/assets/cfe0f4cb-3399-4d7e-95fe-d39859bdc9f1)
![image](https://github.com/user-attachments/assets/8643909a-e3d3-4f88-8ddc-1e1d0c438394)

If that doesn't suffice, please describe in more detail what you want.

With the exception of particle systems, iTwin.js does not provide a way for you to inject your own shader code.

**Reply by peiyanlu** [reply]
like this: Add contours along the edges of the entire model (Translation software may have some deviations in translation)
![777393-20180706144610493-1873024237](https://github.com/user-attachments/assets/207b2407-7185-4d46-9585-879d480a2e7d)

**Reply by pmconne** [reply]
Those are typically called "silhouettes". In your example they are being generated in screen-space. We support [generating silhouettes](https://www.itwinjs.org/reference/core-frontend/rendering/graphicbuilderoptions/#generateedges) as real geometry, so you get more consistent and controllable width.
If this does not answer you question please be more specific about what you're trying to accomplish.

**Reply by peiyanlu** [reply]
> Those are typically called "silhouettes". In your example they are being generated in screen-space. We support [generating silhouettes](https://www.itwinjs.org/reference/core-frontend/rendering/graphicbuilderoptions/#generateedges) as real geometry, so you get more consistent and controllable width. If this does not answer you question please be more specific about what you're trying to accomplish.

Thank you for your reply, but there are still some differences between the current way and what I want. I will try my best to describe it accurately
The `hilite` API in iTwin provides the ability to draw silhouettes for elements, and I have used it to achieve the following effects:

![image](https://github.com/user-attachments/assets/ed6078fd-2918-43c2-8d4d-38cc12f3ae00)
![image](https://github.com/user-attachments/assets/d1cf958c-05ee-43cd-a59d-9592b72103ad)
![image](https://github.com/user-attachments/assets/831468f5-fb01-48b7-8c97-74e9672a7516)

I did it this way:

```ts
  vp.hilite = new Hilite.Settings(ColorDef.fromString('#FF0000'), 0, 0, Hilite.Silhouette.Thick)
  const reader = vp.iModel.createQueryReader(
    'SELECT ECInstanceId FROM bis.GeometricElement3d',
    undefined,
    { rowFormat: QueryRowFormat.UseJsPropertyNames },
  )
  const ids = (await reader.toArray()).map(k => k.id)
  vp.iModel.selectionSet.replace(ids)
```


This is the effect I want, only outlining the edges of the model. The current implementation method has some imperfections: when I click on other places, the effect will be lost due to changes in `selectionSet`. Therefore, I hope to have a stable way to implement it. I tried to understand the implementation method of `hilite`, but found it a bit difficult, so I need your help

**Reply by pmconne** [reply, marked as answer]
Use [emphasis silhouettes](https://www.itwinjs.org/changehistory/1.6.0/). Same effect as hilite but separately configurable and not coupled to the selection set.

**Reply by peiyanlu** [reply]
> Use [emphasis silhouettes](https://www.itwinjs.org/changehistory/1.6.0/). Same effect as hilite but separately configurable and not coupled to the selection set.

Tanks，this is the effect I hope for
